### PR TITLE
deobfuscate_subtitles

### PR DESCRIPTION
**Pull request by sanderjo**
renames loose .srt files so they get the same basename as the largest file in the directory

Closes https://github.com/sabnzbd/sabnzbd/issues/2880

Test .NZBs with short .srt files that will get renamed:

obfuscated mainfile:

https://raw.githubusercontent.com/sanderjo/NZBs/master/Reftestnzb_Subtitles_Obfuscated_Mainfile_405938efd30b.nzb

non_obfuscated mainfile:

https://raw.githubusercontent.com/sanderjo/NZBs/master/Reftestnzb_Some_Subtitles_1234_ebcb671ed6f6.nzb

### CODE REVIEW

File: sabnzbd/postproc.py [hunk 1]
```diff
@@ -574,6 +574,7 @@ def process_job(nzo: NzbObject) -> bool:
                     # Deobfuscate the filenames
                     logging.info("Running deobfuscate")
                     deobfuscate.deobfuscate(nzo, newfiles, nzo.final_name)
+                    deobfuscate.deobfuscate_subtitles(workdir_complete)
```
Comment (Safihre) [inline sabnzbd/postproc.py:577]: I just can't be used like this.
For users that use the `folder/*` notation for their Category settings, all jobs will be unpacked to the same folder without sub-folders for each job. So the `workdir_complete` could contain 10 different movies and assosiated files.
And a lot of users have this, since it's the solution when Googling for "stop SABnzbd from creating job folders".

You will need to change `deobfuscate.deobfuscate` to update the `newfiles` list, and only use that list as your input for `deobfuscate.deobfuscate_subtitles`.

If you want, I can make that change (the updating of `newfiles` in `deobfuscate.deobfuscate`)?

Comment (sanderjo) [reply sabnzbd/postproc.py:577]: Yes, great if you can do that.

**Comment by sanderjo** [comment]
> What happens in case a job contains multiple larger files, each accompanied by already correctly named srt files?

Good point. In the existing main deobfuscate(), there is already a check for that, via `first_file_is_much_bigger(filelist)`. I'll investigate & solve.

Thanks

File: sabnzbd/postproc.py (hunk 1 above)
Comment (Safihre) [reply sabnzbd/postproc.py:577]:
```suggestion
                    deobfuscate.deobfuscate_subtitles(newfiles)
```
You didn't use the output of the previous one?

File: sabnzbd/deobfuscate_filenames.py [hunk 2]
```diff
@@ -225,90 +233,151 @@ def deobfuscate(nzo, filelist: List[str], usefulname: str):
...
+
+    # if directory, create list from that directory
+    if type(input) == str and os.path.isdir(input):
+        all_files = [os.path.join(root, file) for root, dirs, files in os.walk(input) for file in files]
+    elif type(input) == list:
+        # already a list, so nothing to do
+        all_files = input
+    else:
+        logging.debug(f"{input} no directory and no list")
+        return None
```
Comment (Safihre) [inline sabnzbd/deobfuscate_filenames.py:356]: This whole section seems unnecessary, since we always use a list as input.
Just assume that the input is a list and then modify the tests to input a list instead of a directory.

File: sabnzbd/deobfuscate_filenames.py [hunk 3]
```diff
@@ -225,90 +233,151 @@ def deobfuscate(nzo, filelist: List[str], usefulname: str):
...
+        logging.debug("No .srt files found, so nothing to do")
+        return None
+
+    # check there is a clearly biggest file
+    if not (biggest_file := get_biggest_file(all_files)):
+        logging.debug("No clearly biggest file found, so no subtitle renaming feasible")
+        return None
+
+    biggest_file_without_ext = without_extension(biggest_file)  # get full path base name of biggest file
+    logging.debug(f"biggest_file {biggest_file} with base {biggest_file_without_ext}")
```
Comment (Safihre) [inline sabnzbd/deobfuscate_filenames.py:369]: Logging line should be made a bit more human-friendly or be removed.

File: sabnzbd/deobfuscate_filenames.py [hunk 4]
```diff
@@ -225,90 +233,151 @@ def deobfuscate(nzo, filelist: List[str], usefulname: str):
...
+
+    # handle srt files one by one
+    for srt_file in srt_files:
+        if without_extension(srt_file).startswith(biggest_file_without_ext):
+            # already the same start as the biggest file, so skip
+            continue
+        # not the same start, so rename the srt file
+        filename_only = os.path.basename(srt_file)  # like "14_English.srt", so without path
+        filename_only = filename_only.replace("_", ".")  # replace underscore with dot
```
Comment (Safihre) [inline sabnzbd/deobfuscate_filenames.py:378]: Why? This seems unrelated functionality?

File: sabnzbd/deobfuscate_filenames.py [hunk 5]
```diff
@@ -225,90 +233,151 @@ def deobfuscate(nzo, filelist: List[str], usefulname: str):
...
+        # now put that name after the base name of the biggestfile:
+        new_full_name = f"{biggest_file_without_ext}.{filename_only}"  # put (renamed) srt behind that
+        unique_filename = get_unique_filename(new_full_name)  # make sure it's really unique
+        logging.debug(f"Renaming subtitle {srt_file} to {unique_filename}")
```
Comment (Safihre) [inline sabnzbd/deobfuscate_filenames.py:382]: This is already logged by the `renamer` function, so it can be removed.

File: sabnzbd/deobfuscate_filenames.py (hunk 4 above)
Comment (sanderjo) [reply sabnzbd/deobfuscate_filenames.py:378]: It makes VLC and thus the user happy ... which is the goal of this PR.

See https://github.com/sabnzbd/sabnzbd/issues/2880#issuecomment-2168680195 : if we keep the "_" in the filename, it won't show up nicely in VLC. If we replace the "_" with ".", it does appear nicely in VLC.

Comment (sanderjo) [reply sabnzbd/deobfuscate_filenames.py:378]: I've added a comment to make that clearer

**Review by Safihre — CHANGES_REQUESTED** [review]
I think we should also update the description on the Switches page to include this new functionality.
image

File: sabnzbd/deobfuscate_filenames.py [hunk 6]
```diff
@@ -225,90 +232,140 @@ def deobfuscate(nzo, filelist: List[str], usefulname: str):
...
+        if without_extension(srt_file).startswith(biggest_file_without_ext):
+            # already the same start as the biggest file, so skip
+            continue
+        # not the same start, so rename the srt file
+        filename_only = os.path.basename(srt_file)  # like "14_English.srt", so without path
+        filename_only = filename_only.replace("_", ".")  # replace underscore with dot; better for VLC
+        # now put that name after the base name of the biggestfile:
+        new_full_name = f"{biggest_file_without_ext}.{filename_only}"  # put (renamed) srt behind that
+        unique_filename = get_unique_filename(new_full_name)  # make sure it's really unique
+        renamer(srt_file, unique_filename)  # ... and rename actual file on disk
```
Comment (Safihre) [inline sabnzbd/deobfuscate_filenames.py:371]: Should we maybe also include a History update?
```
nzo.set_unpack_info("Deobfuscate", T("Deobfuscate renamed %d subtitle file(s)") % nr_files_renamed)
```

File: sabnzbd/deobfuscate_filenames.py [hunk 7]
```diff
@@ -225,90 +232,140 @@ def deobfuscate(nzo, filelist: List[str], usefulname: str):
...
+    logging.debug(f"Using as base filename: {biggest_file_without_ext}")
+
+    # handle srt files one by one
+    for srt_file in srt_files:
+        if without_extension(srt_file).startswith(biggest_file_without_ext):
+            # already the same start as the biggest file, so skip
+            continue
+        # not the same start, so rename the srt file
+        filename_only = os.path.basename(srt_file)  # like "14_English.srt", so without path
+        filename_only = filename_only.replace("_", ".")  # replace underscore with dot; better for VLC
```
[the remark this answers is not in the document]
Comment (sanderjo) [reply sabnzbd/deobfuscate_filenames.py:367]: I do not agree, but to avoid discussion, and get the PR done, I'll remove the line.

File: sabnzbd/deobfuscate_filenames.py (hunk 6 above)
Comment (sanderjo) [reply sabnzbd/deobfuscate_filenames.py:371]: Good idea. Done
